src/ImageSharp/Formats/Bmp/BmpEncoderCore.cs

```diff
@@ -105,23 +106,23 @@ public void Encode<TPixel>(Image<TPixel> image, Stream stream)
                 yPelsPerMeter: vResolution);
 
             var fileHeader = new BmpFileHeader(
-                type: 19778, // BM
-                fileSize: 54 + infoHeader.ImageSize,
+                type: BmpConstants.TypeMarkers.Bitmap,
+                fileSize: BmpFileHeader.Size + infoHeaderSize + infoHeader.ImageSize,
                 reserved: 0,
-                offset: 54);
+                offset: BmpFileHeader.Size + infoHeaderSize);
 
 #if NETCOREAPP2_1
-            Span<byte> buffer = stackalloc byte[40];
+            Span<byte> buffer = stackalloc byte[infoHeaderSize];
 #else
-            byte[] buffer = new byte[40];
+            byte[] buffer = new byte[infoHeaderSize];
 #endif
             fileHeader.WriteTo(buffer);
 
             stream.Write(buffer, 0, BmpFileHeader.Size);
 
-            infoHeader.WriteTo(buffer);
+            infoHeader.WriteV3Header(buffer);
```

Comment:
How does this affect alpha values. Will decoders now have to check each byte to see if we are the V3 ICO variant? 

It would be really nice if we stored the header type as an enum **Name - Length** in the metadata and gradually build up our ability to write each type starting with v3 and v4.

Reply:
@brianpopow Ah so that was just a method name change. I thought we were changing the header itself!

We already have a method to encode 32bit bitmaps, I'm assuming we're doing that wrong then.

https://github.com/SixLabors/ImageSharp/blob/db0fc0dd7df4b23236120312419f21916b6c6d8a/src/ImageSharp/Formats/Bmp/BmpEncoderCore.cs#L163

By V3 ICO variant I mean this. https://github.com/SixLabors/ImageSharp/issues/732#issuecomment-447151183

Yeah, we are not using the metadata, we certainly should be.

Reply:
yes only the names changed, i wanted to make it clear, that a V3 header is written.

This V3 ICO variant was new to me. I could not find much about it, except the mozilla comment you mentioned. 

My understanding is, that a V3 header implies, no support for alpha channel. As you mentioned in the other ticket, what you called Option 2 could be a way to solve this, but it feels a bit hacky to me.

I think a better way would be to write a V4 header with BITFIELDS compression to indicate which bits are really the alpha channel.

Reply:
I've seen the similar handling code in several decoders now. Chrome, Mozilla, ImageMagick, Gimp, all do this when decoding V3 bitmaps. It is hacky IMO also but a necessary evil. 

I'm all for writing the V4 header! 👍

Reply:
The encoder now writes V4 header. 

I wanted to add rgb32.bmp to the list of testimages for the decoder, but this failed, because of the issue described in #732 that alpha channel is not correctly decoded. This image is a bitmap with a V3 header, but no alpha channel. In the info header it is marked as a 32 bpp image, but the alpha values are all 0, meaning no alpha channel is there. As a consequence the encoder would encode images with v4 header wrong, because it expects there an alpha channel for 32 bpp images.

As a temporary workaround i have ignored the alpha channel during encoding, when its a V3 header. This should be removed once #732 is fixed properly. I have also disabled rgb32.bmp from the decoder tests, until #732 is fixed.